### PR TITLE
net5 compilation and delete obsolete code

### PR DESCRIPTION
**Pull request by SergeiPavlov**
This copy of @AlexUstinov 's PR into http://github.com/servicetitan/dataobjects-net

### CODE REVIEW

**Review by alex-kulakov — CHANGES_REQUESTED** [review]
Please take a look at the comments.
Also I don't agree with ```PostgresqlSqlDml``` rename. The class supposes to contain an extended operations for, in this case, PostgreSQL. Its name is similar to general ```SqlDml``` static class which contains variety of different methods that help building ```SqlExpression```s.  Thought ```PostgresqlSqlDml``` contains only geometric functions it can be extended with more functions. Same thing for ```PostgresqlSqlFunctionType``` rename.

About local functions usage. I noticed several of them that I wanted to change. e.g. in ```Node.cs``` and ```DelegateHelper```. Local functions are faster than Func/Action so I'm all for it. They also can be placed anywhere within parent method as oppose to delegates. So why not to improve code reading experience by moving them, for example, to the end like it is shown in examples [here](https://docs.microsoft.com/ru-ru/dotnet/csharp/programming-guide/classes-and-structs/local-functions)? They are not always should be read, especially when their name is clear and reflects what they do. But having them in the middle of the parent methods body forces reader to find where certain function ends to continue reading main method. What's your arguments for having local functions in the middle of the parent method?

File: Orm/Xtensive.Orm/Sql/Model/PairedNodeCollection.cs [hunk 1]
```diff
@@ -38,6 +39,14 @@ public override void Add(TNode item)
       item.UpdatePairedProperty(property, owner);
     }
 
+    /// <inheritdoc/>
+    public override void AddRange(IEnumerable<TNode> items)
+    {
+      foreach (var item in items) {
+        Add(item);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/Model/PairedNodeCollection.cs:48]: The check is performed in `Add()` method.
I don't think this method is supposed to be atomic

File: Orm/Xtensive.Orm/Orm/Building/Builders/PartialIndexFilterBuilder.cs [hunk 2]
```diff
@@ -71,21 +72,31 @@ protected override Expression VisitMemberAccess(MemberExpression originalMemberA
...
-      var fieldName = StringExtensions.Join(".", memberAccessSequence.Select(item => item.Member.Name));
+      }
+      var nameBuilder = new StringBuilder();
+      for (var i = memberAccessSequence.Count; i-- > 0;) {
+        nameBuilder
+          .Append(memberAccessSequence[i].Member.Name)
+          .Append('.');
+      }
+      nameBuilder.Length -= 1;
+      var fieldName = nameBuilder.ToString();
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Building/Builders/PartialIndexFilterBuilder.cs:95]: Reverse().Join()  makes more allocations for iterators

File: Orm/Xtensive.Orm/Modelling/Node.cs [hunk 3]
```diff
@@ -41,8 +41,8 @@ public abstract class Node : LockableBase,
     public static readonly char PathEscape = '\\';
 
     [NonSerialized]
-    private static ThreadSafeDictionary<Type, PropertyAccessorDictionary> cachedPropertyAccessors = 
-      ThreadSafeDictionary<Type, PropertyAccessorDictionary>.Create(new object());
+    private static ConcurrentDictionary<Type, Lazy<PropertyAccessorDictionary>> cachedPropertyAccessors =
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Modelling/Node.cs]: fixed

File: Orm/Xtensive.Orm.Tests/Upgrade/UpgradeAndNamingRulesAdvanced/TargetModels.cs [hunk 4]
```diff
@@ -4,6 +4,8 @@
 // Created by: Alexey Kulakov
 // Created:    2016.02.03
 
+using System.Collections.Generic;
+
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm.Tests/Upgrade/UpgradeAndNamingRulesAdvanced/TargetModels.cs]: fixed

File: Orm/Xtensive.Orm/Comparison/AdvancedComparer.cs [hunk 5]
```diff
@@ -21,11 +21,11 @@ namespace Xtensive.Comparison
   [Serializable]
   public sealed class AdvancedComparer<T>: MethodCacheBase<IAdvancedComparer<T>>
   {
-    private static ThreadSafeCached<AdvancedComparer<T>> systemCached =
-      ThreadSafeCached<AdvancedComparer<T>>.Create(new object());
+    private static Lazy<AdvancedComparer<T>> systemCached =
+      new Lazy<AdvancedComparer<T>>(() => ComparerProvider.System.GetComparer<T>());
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Comparison/AdvancedComparer.cs]: fixed name violation

File: Orm/Xtensive.Orm/Comparison/AdvancedComparer.cs [hunk 6]
```diff
@@ -21,11 +21,11 @@ namespace Xtensive.Comparison
...
 
-    private static ThreadSafeCached<AdvancedComparer<T>> defaultCached =
-      ThreadSafeCached<AdvancedComparer<T>>.Create(new object());
+    private static Lazy<AdvancedComparer<T>> defaultCached =
+      new Lazy<AdvancedComparer<T>>(() => ComparerProvider.Default.GetComparer<T>());
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Comparison/AdvancedComparer.cs]: Fixed

File: Orm/Xtensive.Orm/Orm/Rse/ColumnGroupCollection.cs [hunk 7]
```diff
@@ -5,65 +5,68 @@
...
-        return this.ElementAt(groupIndex);
-      }
-    }
+    public ColumnGroup this[int groupIndex] => items[groupIndex];
+
+    /// <inheritdoc/>
+    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+    /// <inheritdoc/>
+    public IEnumerator<ColumnGroup> GetEnumerator() => items.GetEnumerator();
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Rse/ColumnGroupCollection.cs]: fixed

File: Orm/Xtensive.Orm/Orm/Building/PrefetchActionContainer.cs [hunk 8]
```diff
@@ -21,14 +19,14 @@ internal sealed class PrefetchActionContainer
   {
     private readonly TypeInfo type;
     private List<AssociationInfo> associations;
-    private ReadOnlyList<PrefetchFieldDescriptor> fields;
+    private ReadOnlyCollection<PrefetchFieldDescriptor> fields;
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Building/PrefetchActionContainer.cs]: correct, fixed

File: Orm/Xtensive.Orm/Orm/Building/PrefetchActionContainer.cs [hunk 9]
```diff
@@ -21,14 +19,14 @@ internal sealed class PrefetchActionContainer
...
 
     public Action<SessionHandler, IEnumerable<Key>> BuildPrefetchAction()
     {
-      fields = new ReadOnlyList<PrefetchFieldDescriptor>(
-        associations
+      fields = associations
           .Select(association => new PrefetchFieldDescriptor(association.OwnerField, true, false))
-          .ToList());
+          .ToList()
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Building/PrefetchActionContainer.cs]: Used `.SelectToList()`

File: Orm/Xtensive.Orm/Orm/Building/Builders/ModelBuilder.cs [hunk 10]
```diff
@@ -26,7 +25,25 @@ internal sealed class ModelBuilder
...
+
+      public override int GetHashCode() => (Name ?? string.Empty).GetHashCode();
+
+      public TypeKey(string name, Type ownerType, Type targetType) {
+        Name = name;
+        OwnerType = ownerType;
+        TargetType = targetType;
+      }
+    }
+    private static ConcurrentDictionary<TypeKey, Lazy<Type>> generatedTypes = new ConcurrentDictionary<TypeKey, Lazy<Type>>();
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Building/Builders/ModelBuilder.cs]: done. Also converted nested function `AuxiliaryTypeFactory` into static field

File: Orm/Xtensive.Orm/Sql/ValueTypeMapping/TypeMapper.cs [hunk 11]
```diff
@@ -269,10 +268,14 @@ public virtual object ReadByteArray(DbDataReader reader, int index)
       if (value == null || value is byte[])
         return value;
 
-      var formatter = new BinaryFormatter();
-      var stream = new MemoryStream();
-      formatter.Serialize(stream, value);
-      return stream.ToArray();
+      throw new NotSupportedException("There is no support of SqlGeometry, SqlGeography, or other complex SQL types to the moment");
+      // As far as SqlGeometry and SqlGeography have no support in .Net 5
+      // we don't need to provide a functionality reading those data as byte arrays
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/ValueTypeMapping/TypeMapper.cs:273]: In my view names `SqlGeometry`, `SqlGeography` are general enough.
SQL server API users know them. But for others they have no relation to Microsoft

File: Orm/Xtensive.Orm/Sql/Model/PairedNodeCollection.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alex-kulakov) [reply Orm/Xtensive.Orm/Sql/Model/PairedNodeCollection.cs:48]: In the base type thought the check exists in both ```AddRange``` and ```Add```. Here, ```AddRange``` is made to ignore the check. This is weird to me, so I asked why it differs

File: Orm/Xtensive.Orm/Sql/ValueTypeMapping/TypeMapper.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (alex-kulakov) [reply Orm/Xtensive.Orm/Sql/ValueTypeMapping/TypeMapper.cs:273]: No, they are from a Microsoft's assembly for MS Sql Server. In DO 5 these types were available to use in Entities, of course with MS Sql Server only. Other providers didn't have such types to use in Entities.

File: Orm/Xtensive.Orm/Sql/Model/PairedNodeCollection.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/Model/PairedNodeCollection.cs]: I see the method `.AddRange()` was added looking at method `override Add()` (which is without locking check).
I agree, explicit check would be more consistent. Added it

**Comment by AlexUstinov** [comment]
> Please take a look at the comments. Also I don't agree with `PostgresqlSqlDml` rename. The class supposes to contain an extended operations for, in this case, PostgreSQL. Its name is similar to general `SqlDml` static class which contains variety of different methods that help building `SqlExpression`s. Thought `PostgresqlSqlDml` contains only geometric functions it can be extended with more functions. Same thing for `PostgresqlSqlFunctionType` rename.

I'm sorry, but I didn't notice the similarity with the `SqlDml` class, but the new name is still similar though. The intention was to provide a meaningful comment to the class and I tried to identify what it actually does. Honestly, we can afford having a dedicated container type for any target DML subdomain like geometry in that case. But I don't mind renaming it back, please just suggest a proper XML comment for the type.

**Comment by alex-kulakov** [comment]
For PostgresqlSqlFunctionType I would suggest 
```csharp
  /// <summary>
  /// Contains additional provider-specific SQL functions.
  /// These functions are not presented in general set of supported SQL
  /// functions and are used only within provider, e.g for operations on
  /// provider-specific types.
  /// </summary>
  public static class PostgresqlSqlFunctionType
```

```PostgresqlSqlDml``` had summary, probably it could be better, maybe like that
```csharp
  /// <summary>
  /// A factory for SQL DML operations for Postgresql DBMS.
  /// Similarly to <see cref="Xtensive.Sql.SqlDml"/> it contains
  /// factory methods for provider-specific <see cref="SqlExpression"/>s.
  /// </summary>
  public static class PostgresqlSqlFunctionType
```

Another type that lacks XML is ```SqlCustomFunctionType```. I suggest following description
```csharp
/// <summary>
  /// Defines custom function type.
  /// Used to define functions that are ouside of standard SQL functions
  /// supported by majority of providers. Such functions should be
  /// handleded by the provider within which they are declared.
  /// </summary>
  [Serializable]
  public sealed class SqlCustomFunctionType : IEquatable<SqlCustomFunctionType>

**Comment by SergeiPavlov** [comment]
Renamed the classes back and changed the XML comment as suggested

**Review by alex-kulakov — APPROVED** [review]

**Comment by alex-kulakov** [comment]
@AlexUstinov and @SergeiPavlov Tests are failing with NRE with stacks similar to this
```
KeyFactory.CreateGenericKey(Domain domain, String nodeId, TypeInfo type, TypeReferenceAccuracy accuracy, Tuple tuple, Int32[] keyIndexes) line 150
KeyFactory.Materialize(Domain domain, String nodeId, TypeInfo type, Tuple value, TypeReferenceAccuracy accuracy, Boolean canCache, Int32[] keyIndexes) line 60
ItemMaterializationContext.Materialize(Int32 entityIndex, Int32 typeIdIndex, TypeInfo type, Pair`1[] entityColumns, Tuple tuple) line 51
lambda_method100(Closure , Object[] , Tuple , ItemMaterializationContext )
<>c__DisplayClass2_0`4.<Bind>b__0(T2 arg2, T3 arg3) line 37
ItemMaterializer`1.Materialize(Tuple tuple, MaterializationContext context, ParameterContext parameterContext) line 24
MaterializingReader`1.get_Current() line 46
Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
Enumerable.First[TSource](IEnumerable`1 source)
QueryResultExtensions.ToScalar[TResult](QueryResult`1& sequence, ResultAccessMethod resultAccessMethod) line 15
TranslatedQuery.ExecuteScalar[TResult](Session session, ParameterContext parameterContext) line 60
QueryProvider.<ExecuteScalar>g__ExecuteScalarQuery|7_0[TResult](TranslatedQuery query, Session session, ParameterContext parameterContext) line 78
QueryProvider.Execute[TResult](Expression expression, Func`4 runQuery) line 107
QueryProvider.ExecuteScalar[TResult](Expression expression) line 81
QueryProvider.Execute[TResult](Expression expression) line 72
Queryable.First[TSource](IQueryable`1 source)
InterfaceTest.MainTest() line 134
```
Didn't tested the changes? And I missed this too so master is broken

**Comment by alex-kulakov** [comment]
I'm reverting merge.

**Comment by SergeiPavlov** [comment]
> I'm reverting merge.

The NRE is fixed

**Pull request merged**